### ISSUE TITLE
accessing metadata

### ISSUE THREAD

**Issue by mr-c** [common-workflow-language/cwljava#10]
See https://github.com/common-workflow-language/workflows/blob/master/tools/samtools-index.cwl

**Comment by portah** [comment]
The other example (when metadata is stored in separate file)
https://github.com/common-workflow-language/workflows/blob/master/tools/samtools-faidx.cwl

**Comment by pgrosu** [comment]
@portah, @tetron, @mr-c - You guys are gonna love this!  Andrey was very nice and gave me a lot of tutorials after our meeting today, which got me very excited.  I realized the power of RDF and Fuseki, which made me wonder if I can run it from the command-line directly - which will later on naturally lead to integrating it into the CWL Java SDK.  So check this out - and this is just from the Windows command-line and without any server running (though we will absolutely need the Fuseki server for other things):

``` Bash
> wget --quiet http://mirrors.ocf.berkeley.edu/apache/jena/binaries/apache-jena-3.0.1.zip

> unzip -q apache-jena-3.0.1.zip

> cd apache-jena-3.0.1

> wget --quiet https://jena.apache.org/tutorials/sparql_data/vc-db-1.rdf

> wget --quiet https://jena.apache.org/tutorials/sparql_data/q1.rq

> cat q1.rq
SELECT ?x
WHERE
 { ?x <http://www.w3.org/2001/vcard-rdf/3.0#FN> "John Smith" }

> pwd
c:\test\apache-jena-3.0.1

> SET JENA_HOME=c:\test\apache-jena-3.0.1

> bat\sparql.bat --data=vc-db-1.rdf --query=q1.rq
---------------------------------
| x                             |
=================================
| <http://somewhere/JohnSmith/> |
---------------------------------
```

Basically I can now query RDF docs, and in theory it should be possible to integrate it as part of the CWL Java SDK, since it is only Jar files.  This way the authorship information that @mr-c would like to have can be stored as a Java Binary Object for reading, writing and querying.  The Fuseki server will also provide other powerful features.

Many thanks!
~p

**Comment by mr-c** [comment]
Great news! :+1:

**Comment by portah** [comment]
Yeap, you can use jena's files and code, However you have to have RDF files, means CWL has to be converted into RDF. And as I understand there is no JAVA code that can do that, yet.

**Comment by pgrosu** [comment]
Hi Andrey,

Sorry to reply so late, but I was looking into this for the past few days.  I noticed that Jena has Java RDF reading-and-writing  capability - but I think we can do better:

https://jena.apache.org/documentation/io/

So opportunity knocks :smile:  So looking through this primer:

https://www.w3.org/TR/2014/NOTE-rdf11-primer-20140624

Then we can easily write a RDF writer as a Vertex->Edge->Vertex function that becomes the `<subject> <predicate> <object>` model you showed me the other day.  This can even be stored as an adjacency matrix.

Any other recommendations are always welcome :)

~p

**Comment by tetron** [comment]
At least initially it might be sufficient to include a "metadata" accessor on the generated objects that is a map holding any extension fields found on the object during load.  That would meet the minimum requirements that metadata is preserved.

**Comment by pgrosu** [comment]
So like a graph object that can be traversed?  If so, that should be fairly easy.
